File: src/main/clients/openrgb/effects/impl/DanceFloor.ts
import Client from '../../client/client';
import Device, { RGBColor } from '../../client/device';
import { HSVColor } from '../../client/utils';
import { AbstractEffect } from '../AbstractEffect';

export class DanceFloor extends AbstractEffect {
  public getName(): string {
    return 'Dance floor';
  }

  private getRandom(length): Array<RGBColor> {
    const random: Array<RGBColor> = [];
    for (let i = 0; i < length; ++i) {
      const hsv = HSVColor(
        Math.random() * 359,
        Math.random() * 0.5 + 0.5,
        Math.random() * 0.5 + 0.5
      );

      const red = Math.floor(hsv.red * this.brightness);
      const green = Math.floor(hsv.green * this.brightness);
      const blue = Math.floor(hsv.blue * this.brightness);
      random.push({ red, green, blue });
    }

    return random;
  }

  protected applyEffect(client: Client, devices: Array<Device>): void {
    const loop = (): void => {
      if (this.isRunning) {
        devices.forEach((dev) => {
          client.updateLeds(dev.deviceId, this.getRandom(dev.colors.length));
        });
        setTimeout(loop, 500);
      } else {
        this.hasFinished = true;
      }
    };

    loop();
  }
}

File: rcc/clients/tcp/openrgb/effects/dance_floor.py
import random

from rcc.clients.tcp.openrgb.client.utils import RGBColor
from rcc.clients.tcp.openrgb.effects.base.abstract_effect import AbstractEffect
from rcc.utils.singleton import singleton


@singleton
class DanceFloor(AbstractEffect):
    """Dance floor effect"""

    def __init__(self):
        super().__init__("Dance floor")

    def _get_random(self, length):
        randoms = []
        for _ in range(length):
            randoms.append(RGBColor.fromHSV(random.randint(0, 359), random.randint(0, 25) + 75, 100))
        return randoms

    def apply_effect(self):
        while self._is_running:
            for dev in self._devices:
                self._set_colors(dev, self._get_random(len(dev.colors)))
            self._sleep(0.5)  # 500 ms


dance_floor = DanceFloor()
